XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
```

Comment:
Thanks Trevor for the link to the doc on the event loop - that's really helpful for me.  If the goal of the API is to "add events to the life cycle of the event loop", then why isn't the event loop's phase exposed through API calls? Now, I'm not saying that it should be part of the API, but given stated goal, it isn't clear to why this is omitted.

Which is one of the things I'm driving at:  Clarification/simplification around the goals API's goals will help get everyone on the same page around the utility of the API & how it is to be used.  Ideally, this would include some definitions, goals/non-goals and some canonical use cases/examples.

> async wrap has always been partially tied to implementation details. that's the reason for its existence, and attempting to abstract that would remove the utility from the user.

Can you expand on your example here?  I'm not following how the description of the API impacts the ability to understand event ordering.

> if we wanted all this abstracted away then we could simply wrap all the js calls. that's not the purpose for async wrap.

Again, precisely what I'm driving for - what is the purpose of the API?  My understanding is the goal is to expose lifecycle events around async code execution.  If not, then that's fine.  If the goal is to provide lifecycle events around handles, then great, but it begs the question of "what is a handle and why does the user care?" (per @jasnell's comment).

> It doesn't work like this. An oncomplete or similar is assigned to the newly created request upon construction. Hence why I chose to notify on construction of the instance

Sorry, I'm not following this. 

> I think what you want is... 

I'm not following.  We may be talking past each other with terminology.  What I need is a notification model around async code execution in node.  That is, I want to know precisely the following:
 - when is a callback _c_ "made available" for invocation, and what is the parent invocation "making it available"?   
 - when does a callback _c_ begin invocation?
 - when does a callback _c_ end invocation?
 - when is a callback _c_ no longer available for invocation (arguably optional, but let's stick with it for now as it cleanly maps to proposed `destroy()` call). 

> As far as measuring fire-to-completion, it's impossible to do reliably in node.

I think this is OK, but again, it will help to explicitly list the goals & non-goals of the API.  At a minimum it will get everyone reading this on the same page.

I'm happy to take a stab at writing up what I think is a user model for the API, and something that fits cleanly into current async wrap implementation.  Likely not 100% correct, but at a minimum useful to highlight differences in understanding.  Let me know if people think this is a useful effort.